New tool: biobambam/bamsormadup

matthdsm opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->

Add biobambam/bamsormadup
fixes #1479

## PR checklist
- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

matthdsm commented:
I'm trying to run the tests locally, but keep running into errors with both docker and conda.
AFAIK everything should be ok, but I suppose I'm still missing something.

Some help would be much appreciated
M

Emiller88 commented:
I ran it with manually with conda and got the error:
`PROFILE=conda nextflow run tests/modules/biobambam/bamsormadup -entry test_biobambam_bamsormadup -c tests/config/nextflow.config`
```
Command error:
  libmaus2::bambam::parallel::GenericInputControlReadWorkPackageDispatcher: Invalid empty stream is invalid bgzf/BAM

  /usr/local/bin/../lib/./libmaus2_stacktrace.so.2(libmaus2::stacktrace::StackTrace::StackTrace()+0x67)[0x7f4181dcce37]
  /usr/local/bin/../lib/libmaus2_exception.so.2(libmaus2::exception::LibMausException::LibMausException()+0x2c)[0x7f418247ea7c]
  bamsormadup(+0x648a1)[0x555ad5f518a1]
  bamsormadup(+0x9a4a8)[0x555ad5f874a8]
  bamsormadup(+0x4a565)[0x555ad5f37565]                                                                      /usr/local/bin/../lib/libstdc++.so.6(+0xcc9d4)[0x7f41823579d4]
  /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3)[0x7f41820a0fa3]
  /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7f4181fd14cf]

  libmaus2::bambam::parallel::GenericInputControlReadWorkPackageDispatcher: Invalid empty stream is invalid bgzf/BAM

  /usr/local/bin/../lib/./libmaus2_stacktrace.so.2(libmaus2::stacktrace::StackTrace::StackTrace()+0x67)[0x7f4181dcce37]
  /usr/local/bin/../lib/libmaus2_exception.so.2(libmaus2::exception::LibMausException::LibMausException()+0x2c)[0x7f418247ea7c]
  bamsormadup(+0x648a1)[0x555ad5f518a1]
  bamsormadup(+0x9a4a8)[0x555ad5f874a8]
  bamsormadup(+0x4a565)[0x555ad5f37565]
  /usr/local/bin/../lib/libstdc++.so.6(+0xcc9d4)[0x7f41823579d4]
  /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3)[0x7f41820a0fa3]
  /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7f4181fd14cf]
```

jfy133 commented on modules/biobambam/bamsormadup/main.nf:
```diff
@@ -0,0 +1,38 @@
+        'https://depot.galaxyproject.org/singularity/biobambam:2.0.183--h9f5acd7_1':
+        'quay.io/biocontainers/biobambam:2.0.183--h9f5acd7_1' }"
+
+    input:
+    tuple val(meta), path(bam)
+
+    output:
+    tuple val(meta), path("*.bam")        , emit: bam
+    tuple val(meta), path("*.metrics.txt"), emit: metrics
+    path "versions.yml"                   , emit: versions
```
Ok, the help message is a bit difficult to understand for this tool.

Please make sure that all possible input and output channels are listed here [^1]. For example I guess `indexfilename` and `reference` could be possibles?

Optional input channels can be passed `[]` if not needed by the pipeline developer (andin tests)
For optional output files, you can specify `,optional: true` before the `emit`. 

Don't forget to update `meta.yml` accordingly

[^1]: https://nf-co.re/developers/modules#inputoutput-options points 1-3

matthdsm replied:
Do we want to add all of the options? Or can we make some assumptions and leave out some stuff?
If we want everything we'd have to make sure there's support for:
- using a real TMP dir (e.g. scratch) for TMP, instead of the PWD
- add support for a digest with a completely variable filename
- add support form cram output (in conjunction with the reference param)

jfy133 replied:
Basically we don't want to make (in most cases) assumptions what someone would want to do for a given tool. So maybe they want the extra files that the tool needs.

1. In this case it's OK to ignore - the nextflow `work/` is technically a scratch, and presumably(?) a user would never want to access these scratch files?
2. Not sure what you mean here
3. Yes, correct - `    tuple val(meta), path("*.{bam,cram}")        , emit: bam` - there are other modules you can look at to find how this is typically implemented

matthdsm replied:
1. Ok, that checks out.
2. Bamsormadup can output a file checksum with different algorithms, the name for this file can be completely random. How should we handle this? 
3. Ok, should be fixed

jfy133 replied:
> Bamsormadup can output a file checksum with different algorithms, the name for this file can be completely random. How should we handle this?

Ooff, ok that is a tricky one... I guess we can sweep that under the rug if it's completely random...

matthdsm replied:
consider it swept! I don't think that feature sees much use TBH

jfy133 commented on tests/modules/biobambam/bamsormadup/test.yml:
```diff
@@ -0,0 +1,12 @@
+    - path: output/biobambam/test.metrics.txt
+      md5sum: 30d6e7d90bb5df46329d4bc0144ce927
+    - path: output/biobambam/versions.yml
+      md5sum: 0d6f3137ed4515333d73c779f2c24445
```
```suggestion
```

Apparently the general policy is not to include the MD5 sum anymore for versions (not yet documented) because makes it annoying when auto-updating the versions of the tools

jfy133 reviewed (approved):
overall pretty good, just a few requests/suggestions

jfy133 commented on modules/biobambam/bamsormadup/main.nf:
```diff
@@ -0,0 +1,38 @@
+        $args \\
+        I=$bam \\
+        O=${prefix}.bam \\
+        M=${prefix}.metrics.txt \\
```
Please remove `metrics`, file naming should be up to the pipeline developer [^1]

[^1]: https://nf-co.re/developers/modules#naming-conventions point 5

matthdsm replied:
how would you suggest differentiating the filename? Or do you mean remove `M=xxx` and moving it to `$args`?

jfy133 replied:
Just I mean just specify as `{prefix}.txt`, unless the tool can also produce other `.txt` files? Otherwise `.txt` should be enough to differentiate - and a pipeline developer can add `metrics` themselves if they want.

matthdsm replied:
Allright, thanks. This is resolved than.

jfy133 replied:
I see you reverted back to `metrics`, is that required by the tool in the end?

matthdsm commented:
Awesome, thanks for the review. I'll get to it right away!

jfy133 reviewed (changes requested):
Oops something went funky in last review and missed a bunch of comments..

Also tests were failing previously

matthdsm commented:
Is there a common way to check the args for errors?
e.g. If the output is a cram file, and `reference` is not defined we should get an error.

Also, is there a way to set the output extension based on a (possibly undefined) arg (`outputformat`)  in $args?

jfy133 commented:
> Is there a common way to check the args for errors? e.g. If the output is a cram file, and `reference` is not defined we should get an error.
> 
> Also, is there a way to set the output extension based on a (possibly undefined) arg (`outputformat`) in $args?

It's generally OK to add some checks in the `script` block with the other process variable definitions. You can see an example of an error check at the bottom of point 5 here: https://nf-co.re/developers/modules#naming-conventions.

and like wise for the variable extension, you are welcome to write a internal conditional variable to adust this

matthdsm commented:
Most comments are fixed! Now we just need to get the tests to work...

jfy133 commented on modules/biobambam/bamsormadup/main.nf:
```diff
@@ -0,0 +1,44 @@
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def suffix = "cram" if args.contains("outputformat=cram") else "bam"
+
+    if (args.contains("outputformat=cram") and reference == null) error "Reference required for CRAM output."
```
```suggestion
    if (args.contains("outputformat=cram") && reference == null) error "Reference required for CRAM output."
```

Normally the && operator is used AFAIK

matthdsm replied:
My bad. Not used to groovy yet I suppose 😅

jfy133 replied:
Doens't actually matter, I'm just so used to it it caught my eye!

jfy133 commented:
For the failng tests, have you tried running locally?

https://nf-co.re/developers/modules#running-tests-manually

matthdsm commented:
Yeah, but I keep having some issues with the `versions.yml` file it seems:
```
 NF_CORE_MODULES_TEST=1 TMPDIR=~ PROFILE=docker pytest --tag biobambam/bamsormadup --symlink --keep-workflow-wd
============================================================= test session starts ==============================================================
platform darwin -- Python 3.10.4, pytest-7.1.1, pluggy-0.12.0
rootdir: /Users/matdsmet/Projects/nf-core-modules, configfile: pytest.ini
plugins: datafiles-2.0, workflow-1.6.0, cov-3.0.0
collecting ... 
collected 732 items                                                                                                                            

biobambam/bamsormadup test_biobambam_bamsormadup:
        command:   nextflow run tests/modules/biobambam/bamsormadup -entry test_biobambam_bamsormadup -c tests/config/nextflow.config
        directory: /Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup
        stdout:    /Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/log.out
        stderr:    /Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/log.err
'biobambam/bamsormadup test_biobambam_bamsormadup' done.

tests/test_versions_yml.py sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 14%]
sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssFssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 33%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 52%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 70%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 89%]
sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss                                                              [ 99%]
tests/modules/biobambam/bamsormadup/test.yml FFFF                                                                                        [100%] Keeping temporary directories and logs. Use '--kwd' or '--keep-workflow-wd' to disable this behaviour.


=================================================================== FAILURES ===================================================================
_______________________________ test_ensure_valid_version_yml[biobambam/bamsormadup test_biobambam_bamsormadup] ________________________________

workflow_dir = PosixPath('/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup')

    @pytest.mark.workflow(*_get_workflow_names())
    def test_ensure_valid_version_yml(workflow_dir):
        workflow_dir = Path(workflow_dir)
        software_name = workflow_dir.name.split("_")[0].lower()
        try:
            versions_yml_file = workflow_dir / f"output/{software_name}/versions.yml"
>           versions_yml = versions_yml_file.read_text()

tests/test_versions_yml.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = PosixPath('/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/output/bamsormadup/versions.yml')
encoding = 'locale', errors = None

    def read_text(self, encoding=None, errors=None):
        """
        Open the file in text mode, read it, and close the file.
        """
        encoding = io.text_encoding(encoding)
>       with self.open(mode='r', encoding=encoding, errors=errors) as f:

../../miniconda3/lib/python3.10/pathlib.py:1132: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = PosixPath('/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/output/bamsormadup/versions.yml')
mode = 'r', buffering = -1, encoding = 'locale', errors = None, newline = None

    def open(self, mode='r', buffering=-1, encoding=None,
             errors=None, newline=None):
        """
        Open the file pointed by this path and return a file object, as
        the built-in open() function does.
        """
        if "b" not in mode:
            encoding = io.text_encoding(encoding)
>       return self._accessor.open(self, mode, buffering, encoding, errors,
                                   newline)
E       FileNotFoundError: [Errno 2] No such file or directory: '/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/output/bamsormadup/versions.yml'

../../miniconda3/lib/python3.10/pathlib.py:1117: FileNotFoundError

During handling of the above exception, another exception occurred:

workflow_dir = PosixPath('/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup')

    @pytest.mark.workflow(*_get_workflow_names())
    def test_ensure_valid_version_yml(workflow_dir):
        workflow_dir = Path(workflow_dir)
        software_name = workflow_dir.name.split("_")[0].lower()
        try:
            versions_yml_file = workflow_dir / f"output/{software_name}/versions.yml"
            versions_yml = versions_yml_file.read_text()
        except FileNotFoundError:
>           raise AssertionError(
                dedent(
                    f"""\
                    `versions.yml` not found in the output directory.
                    Expected path: `{versions_yml_file}`
    
                    This can have multiple reasons:
                    * The test-workflow failed before a `versions.yml` could be generated.
                    * The workflow name in `test.yml` does not start with the tool name.
                    """
                )
            )
E           AssertionError: `versions.yml` not found in the output directory.
E           Expected path: `/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/output/bamsormadup/versions.yml`
E           
E           This can have multiple reasons:
E           * The test-workflow failed before a `versions.yml` could be generated.
E           * The workflow name in `test.yml` does not start with the tool name.

tests/test_versions_yml.py:33: AssertionError
_________________________________________________________________ test session _________________________________________________________________
'/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/output/biobambam/test.bam' does not exist while it should
_________________________________________________________________ test session _________________________________________________________________
'/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/output/biobambam/test.metrics.txt' does not exist while it should
_________________________________________________________________ test session _________________________________________________________________
'/Users/matdsmet/pytest_workflow_gqwmex1a/biobambam/bamsormadup_test_biobambam_bamsormadup/output/biobambam/versions.yml' does not exist while it should
_________________________________________________________________ test session _________________________________________________________________
'biobambam/bamsormadup test_biobambam_bamsormadup' exited with exit code '1' instead of '0'.
=============================================================== warnings summary ===============================================================
../../miniconda3/lib/python3.10/site-packages/pytest_workflow/plugin.py:391
  /Users/matdsmet/miniconda3/lib/python3.10/site-packages/pytest_workflow/plugin.py:391: UserWarning: .git dir detected: /Users/matdsmet/Projects/nf-core-modules/.git. pytest-workflow will copy the entire .git directory and all files ignored by git. It is recommended to use the --git-aware option.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================================== short test summary info ============================================================
FAILED tests/test_versions_yml.py::test_ensure_valid_version_yml[biobambam/bamsormadup test_biobambam_bamsormadup] - AssertionError: `version...
FAILED tests/modules/biobambam/bamsormadup/test.yml::biobambam/bamsormadup test_biobambam_bamsormadup::output/biobambam/test.bam::should exist
FAILED tests/modules/biobambam/bamsormadup/test.yml::biobambam/bamsormadup test_biobambam_bamsormadup::output/biobambam/test.metrics.txt::should exist
FAILED tests/modules/biobambam/bamsormadup/test.yml::biobambam/bamsormadup test_biobambam_bamsormadup::output/biobambam/versions.yml::should exist
FAILED tests/modules/biobambam/bamsormadup/test.yml::biobambam/bamsormadup test_biobambam_bamsormadup::exit code should be 0
================================================== 5 failed, 727 skipped, 1 warning in 11.81s ==================================================
```

jfy133 commented:
> Yeah, but I keep having some issues with the `versions.yml` file it seems:


This is an unfortunate misdirection , you need to check the actual working directory or the .nexflow.log of the test to see what the error is (I hate python traceback sometimes ;) ) . I suspect the process is failing too early so it doesn't generate the versions.yml and but for whatever reason that specific check is the one that makes the largest error...

matthdsm commented:
Ok, looks like a syntax error,  altough I'm not really sure what's wrong...
```
N E X T F L O W  ~  version 21.10.6
Launching `tests/modules/biobambam/bamsormadup/main.nf` [berserk_meucci] - revision: 1a1ee0d1f7
Module compilation error
- file : /Users/matdsmet/pytest_workflow_pfgis0rq/biobambam/bamsormadup_test_biobambam_bamsormadup/tests/modules/biobambam/bamsormadup/../../../../modules/biobambam/bamsormadup/main.nf
- cause: Unexpected input: '{' @ line 1, column 31.
   process BIOBAMBAM_BAMSORMADUP {
                                 ^

1 error
```

matthdsm commented:
So I tried a bunch of things, but for some reason the syntax doesn't check out....
Any ideas? I'm lost...

jfy133 commented:
> So I tried a bunch of things, but for some reason the syntax doesn't check out.... Any ideas? I'm lost...

Noramlly a missing `,` or `'` or bracket somewhere... sorry can't look more closely right at the moment

matthdsm commented:
Ok, so I was able to fix the syntax error (some whitespace in the conditionals) but now the original errors are back
```
  WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
  libmaus2::bambam::parallel::GenericInputControlReadWorkPackageDispatcher: Invalid empty stream is invalid bgzf/BAM

```

The platform warning is because I'm running the tests on an ARM macbook. The other error seems to say the input data is corrupt

jfy133 commented:
Yup same:

```
Command error:
  libmaus2::bambam::parallel::GenericInputControlReadWorkPackageDispatcher: Invalid empty stream is invalid bgzf/BAM
  
  /usr/local/bin/../lib/./libmaus2_stacktrace.so.2(libmaus2::stacktrace::StackTrace::StackTrace()+0x67)[0x7fdb2e46ee37]
  /usr/local/bin/../lib/libmaus2_exception.so.2(libmaus2::exception::LibMausException::LibMausException()+0x2c)[0x7fdb2eb20a7c]
  bamsormadup(+0x648a1)[0x56270bcfd8a1]
  bamsormadup(+0x9a4a8)[0x56270bd334a8]
  bamsormadup(+0x4a565)[0x56270bce3565]
  /usr/local/bin/../lib/libstdc++.so.6(+0xcc9d4)[0x7fdb2e9f99d4]
  /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3)[0x7fdb2e742fa3]
  /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7fdb2e6734cf]
  
  libmaus2::bambam::parallel::GenericInputControlReadWorkPackageDispatcher: Invalid empty stream is invalid bgzf/BAM
  
  /usr/local/bin/../lib/./libmaus2_stacktrace.so.2(libmaus2::stacktrace::StackTrace::StackTrace()+0x67)[0x7fdb2e46ee37]
  /usr/local/bin/../lib/libmaus2_exception.so.2(libmaus2::exception::LibMausException::LibMausException()+0x2c)[0x7fdb2eb20a7c]
  bamsormadup(+0x648a1)[0x56270bcfd8a1]
  bamsormadup(+0x9a4a8)[0x56270bd334a8]
  bamsormadup(+0x4a565)[0x56270bce3565]
  /usr/local/bin/../lib/libstdc++.so.6(+0xcc9d4)[0x7fdb2e9f99d4]
  /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3)[0x7fdb2e742fa3]
  /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7fdb2e6734cf]

```

Is that also what you get?

I have no idea - when I go into the `work/` directory of a manual test, and run the command manually using `conda` installed biobambam, the command just hangs with no error... just sits there doing nothing. Can you confirm that the conda installed version (at least) works on some of your own data? Maybe the conda recipe is borked?

matthdsm commented:
We use the same version, installed by conda in our production environments, so I'm sure it works. 

Very weird. I'll look into it.

jfy133 commented:
The other option is to pick other BAM files from the nf-core test/datasets and see if the ywork! There are quite a few now.

matthdsm commented:
I've tried using  `params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam']`, but still no luck. `samtools view` of the test data seems to yield a valid result, so it must be something inside the docker image

matthdsm commented:
The issue must be somewhere in the module, because when I update `biobambam/bammarkduplicates` to the latest version (so same docker image), the tests do work.

matthdsm commented:
Downgrading the version of `biobambam/bamsormadup` doesn't help either

jfy133 commented:
@matthdsm if you go into the work directory of the process (after a failing test), and try manually running `bash .command.sh` does it work for you? Can be on any of docker/singularity/conda

matthdsm commented:
I can't run any of the commands locally, the conda recipe isn't built for ARM, so not available on my computer

jfy133 commented:
Ah I see... will need to get more input else where, I thikn it best to post the problem on slack and get some other people to try too... I still can't get it to run. Just hangs...

matthdsm commented:
Turns out I'm actually an idiot and got the command line all wrong 🤦🏻  fixing right now

matthdsm commented:
So I added the `bamcat` command in front of bamsormadup, since bamsormadup takes it's input from stdin.
`bamcat` is part of the biobambam suite, so actually the same tool, which is why I could do this without issue.

matthdsm commented:
`bamcat` can take multiple inputs, making this chain also suitable to merge several bam files while simultaneously sorting/markdup.

I'm still new to nextflow. How would I define multiple bam inputs? I imagine would then be able to do something like `"-I ".join([bam1,bam2,...])`

jfy133 commented on modules/biobambam/bamsormadup/meta.yml:
```diff
@@ -0,0 +1,52 @@
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - bams:
+      type: files
```
```suggestion
      type: file
```

matthdsm replied:
I added `files` because some other tools did so to designate a list of possibly multiple inputs. Changed it!

jfy133 commented on modules/biobambam/bamsormadup/meta.yml:
```diff
@@ -0,0 +1,52 @@
+      type: files
+      description: List containing 1 or more bam files
+  - fasta:
+      type: optional file
+      description: Reference genome
```
```suggestion
      type: file
      description: Reference genome in FASTA format (optional)
```

jfy133 commented on tests/modules/biobambam/bamsormadup/test.yml:
```diff
@@ -0,0 +1,12 @@
+    - path: output/biobambam/test.metrics.txt
+      md5sum: 1721879bea1f3888ecd33b35e6ee0e72
+    - path: output/biobambam/versions.yml
+      md5sum: 841f96891b41daa43faf2072c08419f5
```
```suggestion
```

We are starting to remove this as it makes automated version bumping of modules more difficult

matthdsm commented:
> I see you reverted back to metrics, is that required by the tool in the end?

Not required, but I noticed some other tools used for marking duplicates (picard, bammarkduplicates2) use the same convention, so I decided to stick with it for the sake of continuity

matthdsm commented:
Feel free to merge I everything looks ok to you!

jfy133 commented:
> > I see you reverted back to metrics, is that required by the tool in the end?
> 
> Not required, but I noticed some other tools used for marking duplicates (picard, bammarkduplicates2) use the same convention, so I decided to stick with it for the sake of continuity

Oh really? they were early tools so maybe that was before the guidlines were confirmed...

 Hm ok. Fair enough then!

matthdsm commented:
Thanks for the help and review. Lots more to come!
